javalib/src/main/scala/java/lang/dec2flt/FloatingPointFormat.scala

```diff
@@ -0,0 +1,154 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package java.lang.dec2flt
+
+import java.lang.Math
+import java.math.BigInteger
+
+sealed trait FloatingPointFormat {
+  type Repr
+
+  val ExpBits: Int
+  val SigBits: Int
+  val CeilLog5OfMaxSig: Int
+  val MaxSig: Long
+  val PositiveInfinity: Repr
+
+  final val ExplicitSigBits: Int = SigBits - 1
+
+  def powerOfTen(i: Int): Repr
+  def prevFloat(v: Repr): Repr
+  def nextFloat(v: Repr): Repr
+  def mul(a: Repr, b: Repr): Repr
+  def div(a: Repr, b: Repr): Repr
+  def frexp(v: Repr): (Long, Int)
+  def fromLong(v: Long): Repr
+  def reinterpretBits(v: BigInteger): Repr
+  def toIEEE754(v: FloatingPoint): Repr
+}
+
+object Binary32 extends FloatingPointFormat {
+  type Repr = Float
+
+  final val ExpBits = 8
+  // final val ExplicitSigBits = 23
+  final val SigBits = 24
+
+  final val MaxSig: Long = (1L << SigBits) - 1;
+  final val CeilLog5OfMaxSig = 11
+  final val PositiveInfinity: Repr = java.lang.Float.POSITIVE_INFINITY
+  private final val PowerOfTens = Array(
+    1.0f,
+    10.0f,
+    100.0f,
+    1000.0f,
+    10000.0f,
+    100000.0f,
+    1000000.0f,
+    10000000.0f,
+    100000000.0f,
+    1000000000.0f,
+    10000000000.0f,
+  )
+
+  def powerOfTen(i: Int): Repr = PowerOfTens(i)
+  def prevFloat(v: Repr): Repr = Math.nextDown(v)
+  def nextFloat(v: Repr): Repr = Math.nextUp(v)
+  def mul(a: Repr, b: Repr): Repr = a * b
+  def div(a: Repr, b: Repr): Repr = a / b
+  def frexp(v: Repr): (Long, Int) = {
+    val bits = java.lang.Float.floatToRawIntBits(v)
+    val m = (bits & 0x7fffff + (1 << ExplicitSigBits)).toLong
+    val exp = ((bits >>> ExplicitSigBits) & 0xff) -
+        ((1 << (ExpBits - 1)) - 1) - ExplicitSigBits
```

Comment:
This reads really strangely. Keep it on one line:
```suggestion
    val exp = ((bits >>> ExplicitSigBits) & 0xff) - ((1 << (ExpBits - 1)) - 1) - ExplicitSigBits
```